core/src/serialize/mod.rs

```diff
@@ -329,66 +436,95 @@ pub fn validate(format: ExportFormat, t: &RichTerm) -> Result<(), ExportError> {
+            }
+            ValueContentRef::Bool(_)
+            | ValueContentRef::Record(Container::Empty)
+            | ValueContentRef::Array(Container::Empty)
+            | ValueContentRef::String(_) => Ok(()),
+            ValueContentRef::EnumVariant(EnumVariantData { arg: None, .. }) => Ok(()),
+            ValueContentRef::EnumVariant(EnumVariantData { arg: Some(_), .. }) => {
+                Err(ExportErrorData::NonSerializable(value.clone()).into())
+            }
+            ValueContentRef::Number(n) => {
                 if *n >= *NUMBER_MIN && *n <= *NUMBER_MAX {
                     Ok(())
                 } else {
                     Err(ExportErrorData::NumberOutOfRange {
-                        term: t.clone(),
+                        term: value.clone(),
                         value: n.clone(),
                     }
                     .into())
                 }
             }
-            Record(record) => {
+            ValueContentRef::Record(Container::Alloc(record)) => {
                 record.iter_serializable().try_for_each(|binding| {
                     // unwrap(): terms must be fully evaluated before being validated for
                     // serialization. Otherwise, it's an internal error.
-                    let (id, rt) = binding.unwrap_or_else(|err| {
+                    let (id, value) = binding.unwrap_or_else(|err| {
                         panic!(
                             "encountered field without definition `{}` \
                             during pre-serialization validation",
                             err.id
                         )
                     });
 
-                    do_validate(format, rt)
-                        .map_err(|err| with_elem(err, NickelPointerElem::Field(id)))
+                    do_validate(format, value)
+                        .map_err(|err| err.with_elem(NickelPointerElem::Field(id)))
                 })?;
                 Ok(())
             }
-            Array(array, _) => {
-                array.iter().enumerate().try_for_each(|(index, t)| {
-                    do_validate(format, t)
-                        .map_err(|err| with_elem(err, NickelPointerElem::Index(index)))
-                })?;
+            ValueContentRef::Array(Container::Alloc(array_data)) => {
+                array_data
+                    .array
+                    .iter()
+                    .enumerate()
+                    .try_for_each(|(index, val)| {
+                        do_validate(format, val)
+                            .map_err(|err| err.with_elem(NickelPointerElem::Index(index)))
+                    })?;
                 Ok(())
             }
-            _ => Err(ExportErrorData::NonSerializable(t.clone()).into()),
+            // Not sure if we should allow this. But supporting wrapped values might alleviate the
+            // pre-processing substitution work to be done upfront.
```

Comment:
Given that you're planning to get rid of `Term::Value`, maybe this should just be an error?

Reply:
I agree, but I would say let's do it as part of removal of `Term::Value`, for separation of concerns.Since there's a use here, the compiler will complain and the comment should make it clear what to do, hopefully.